src/element/scale.ts

```diff
@@ -0,0 +1,53 @@
+import { isTextElement } from "./typeChecks";
+import { redrawTextBoundingBox } from "./textElement";
+import { ExcalidrawElement } from "./types";
+
+export function getScaledElement(
+  element: ExcalidrawElement,
+  scale: number,
+): ExcalidrawElement {
+  switch (element.type) {
+    case "selection": {
+      return element;
+    }
+    case "rectangle":
+    case "diamond":
+    case "ellipse":
+    case "line": {
+      const scaledElement = { ...element };
+      scaledElement.x *= scale;
+      scaledElement.y *= scale;
+      scaledElement.width *= scale;
+      scaledElement.height *= scale;
+      scaledElement.strokeWidth *= scale;
+      return scaledElement;
+    }
+    case "arrow": {
+      const scaledElement = { ...element };
+      scaledElement.x *= scale;
+      scaledElement.y *= scale;
+      scaledElement.width *= scale;
+      scaledElement.height *= scale;
+      scaledElement.strokeWidth *= scale;
+      scaledElement.points = scaledElement.points.map(([x, y]) => [
+        x * scale,
+        y * scale,
+      ]);
+      return scaledElement;
+    }
+    default: {
+      if (isTextElement(element)) {
+        const scaledElement = { ...element };
+        scaledElement.x *= scale;
+        scaledElement.y *= scale;
+        const fontSize = parseFloat(scaledElement.font);
```

Comment:
This is legacy from how we handle the `font` property. We should split font size and font family in different properties or have utils that allow us to apply transformations on this string.

I copy pasted code found elsewhere in the codebase to be consistent at the moment.

Reply:
I’m not sure that changing the font size like this is going to work. You may not get the right size since font sizes are discrete numbers.

Have you considered drawing the normal size and scaling those commands using canvas instead?

Reply:
That was my first thought, but in fact it's working 🤔 We could maybe round values to get integers instead, but I think the approach is right ?

Using a temp canvas, scaling it and then copying to the main canvas would kind of go against the approach I took here of not blindy using `scale`.

Reply:
What you can do with canvas is to change the scale on the context before you draw, draw it and revert it back.

I believe you should do that for each draw operation and is the right strategy.

Rather than doing one scale at the beginning, and one scale at the end.

Does it make sense? (This also the answer to your question all the way below)

Reply:
All right. So you are suggesting to remove the `getElementWithZoomScale` function and use canvas' context scale instead ? I originally chose this solution because I wanted more granular control on how zoom was affecting every element.

Reply:
I would translate the mouse coordinates to canvas coordinates before sending it to hit testing. So that it’s also transparent, without having to physically update the elements coordinates.

Reply:
I'm not sure I like that. What I mean by "transparent" is that with my current approach I was aiming for physically modifying elements (with `getElementWithZoomScale`) and for modifying movements deltas (yet to be implemented). This allows:

- Confidence about the elements positions in the canvas at any given moment in time. Today it's zoom, tomorrow it's also rotation.
- We have fine grained control on how zoom affects every type of element (not only by a scale factor). For example, we might want to reduce arrows' number of points to improve performance / preview at 200% zoom.
- 1:1 matching in drawing and logic around canvas elements. This reinforces a single mental model of how things work and allows easier manipulation when mouse coordinates affects elements that are not scaled by the canvas (for example scrollbars or resize handlers).

Reply:
> For example, we might want to reduce arrows' number of points to improve performance / preview at 200% zoom.

This is very interesting, thanks for bringing this up. It made me update my mental model of the problem :)

-------------

In my past experience, every time I tried to transform the objects rather than the environment it ended up beating me in some way. So I've generally avoided doing so. Now, this is something as a feeling and likely not going to convince you ;)

Here are some concrete issues I theorize will happen if we do it:
- Rough has different heuristics as to how to render the elements based on size. So it will be the case that as you zoom in and out, the rendering of the elements will look different, which is going to be jarring and unexpected. We also have one for the size of the arrow cap.
- In a similar vein, fonts can have different rendering for different sizes. On the upside, this makes it more legible but on the downside, it will therefore look different. Also, the font sizes are usually discrete and will make jumps as you do a smooth zoom and potentially be misaligned. We'll also have to do more work to properly place the element when it doesn't maintain the same ratio compared to the rest of the scene.
- I've had issues in the past with drawing lines at subpixel widths. I don't know how canvas deal with it but it's likely that if you ask to draw a 0.9 pixel line it's going to draw a 1px one. Applying a transform will render it as 1px and then do image manipulation to make it 0.9 pixel, the effect you want.
- I was hoping that with #655 we'd be able to reuse the flat images to accelerate rendering at smaller sizes. We wouldn't be able to use the larger image to render the smaller scene if your strategy is to redraw the image with different visual output.

Here are some other things that I'm thinking about related to the code:
- In terms of performance, anytime we're doing a full copy of the scene we lose. This means that we are allocating new memory which is super expensive. I know that we're doing .map() all over the codebase right now and it's bad. To have good performance (when all the low hanging fruit is gone), we want to only have virtual transformations (additions & multiplication) when we operate on the elements one by one as they are going through the rendering phase.

I probably should remove my transparence argument. All the transforms we're adding affect the entire system. This is why I wanted to figure out the value proposition for rotation as I knew it was going to be expensive to implement and maintain.

In any case, if you believe that updating the elements is the right thing to do, please go ahead. I'd love to be proven wrong :)

Reply:
Thanks for the in deep thoughts. I will give it a try tonight, but I think your intuition is more accurate than mine 😁